gapii/cc/pack_encoder.cpp

```diff
@@ -159,13 +150,15 @@ uint64_t PackEncoderImpl::writeTypeIfNew(const Descriptor* desc) {
     }
 
     if (!type_id_cache) {
+        std::lock_guard<std::mutex> lock(mShared->mutex);
         return writeTypeIfNewBlocking(desc);
     }
 
     uint64_t type_id;
 
     auto iter_id = type_id_cache->type_ids.find(desc);
     if (iter_id == end(type_id_cache->type_ids)) {
+        std::lock_guard<std::mutex> lock(mShared->mutex);
```

Comment:
Any reason you move the lock outside of `writeTypeIfNewBlocking`?
It just seems more likely now that someone might call it without a lock.

Reply:
The lock is non-recursive, I can change it to recursive one, which is slightly bigger diff.

Reply:
Fair enough.